cleanup/cleanDockerImages/cleanDockerImages.groovy

```diff
@@ -0,0 +1,135 @@
+    }
+}
+
+def buildParentRepoPaths(path, dryRun, daysToKeep) {
+    def deleted = [], oldSet = [], imagesPathMap = [:], imagesCount = [:]
+    def parentInfo = repositories.getItemInfo(path)
+    simpleTraverse(parentInfo, oldSet, imagesPathMap, imagesCount, daysToKeep)
+    for (img in oldSet) {
+        deleted << img.id
+        if (!dryRun) repositories.delete(img)
+    }
+    for (key in imagesPathMap.keySet()) {
+        def repoList = imagesPathMap[key]
+        def maxImagesCount = imagesCount[key]
+        // If number of current docker images is more than maxcount, delete them
+        if (maxImagesCount <= 0 || repoList.size() <= maxImagesCount) continue
+        repoList = repoList.sort { it[1] }
+        def deleteCount = repoList.size() - maxImagesCount
+        for (i = 0; i < deleteCount; i += 1) {
+            deleted << repoList[i][0].id
+            if (!dryRun) repositories.delete(repoList[i][0])
+        }
+    }
+    return deleted
+}
+
+// Traverse through the docker repo (directories and sub-directories) and:
+// - delete the images immediately if the maxDays policy applies
+// - Aggregate the images that qualify for maxCount policy (to get deleted in
+//   the execution closure)
+def simpleTraverse(parentInfo, oldSet, imagesPathMap, imagesCount, daysToKeep) {
+    def maxCount = null
+    def parentRepoPath = parentInfo.repoPath
+    for (childItem in repositories.getChildren(parentRepoPath)) {
+        def currentPath = childItem.repoPath
+        if (childItem.isFolder()) {
+            simpleTraverse(childItem, oldSet, imagesPathMap, imagesCount, daysToKeep)
+            continue
+        }
+        log.debug("Scanning File: $currentPath.name")
+        if (currentPath.name != "manifest.json") continue
+        // get the properties here and delete based on policies:
+        // - implement daysPassed policy first and delete the images that
+        //   qualify
+        // - aggregate the image info to group by image and sort by create
+        //   date for maxCount policy
+        if (checkDaysPassedForDelete(childItem, daysToKeep)) {
+            log.debug("Adding to OLD MAP: $parentRepoPath")
+            oldSet << parentRepoPath
+        } else if ((maxCount = getMaxCountForDelete(childItem)) > 0) {
+            log.debug("Adding to IMAGES MAP: $parentRepoPath")
+            def parentCreatedDate = parentInfo.created
+            def parentId = parentRepoPath.parent.id
+            def oldmax = maxCount
+            if (parentId in imagesCount) oldmax = imagesCount[parentId]
+            imagesCount[parentId] = maxCount > oldmax ? maxCount : oldmax
+            if (!imagesPathMap.containsKey(parentId)) {
+                imagesPathMap[parentId] = []
+            }
+            imagesPathMap[parentId] << [parentRepoPath, childItem.created]
+        }
+        break
+    }
+}
+
+// This method checks if the docker image's manifest has the property
+// "com.jfrog.artifactory.retention.maxDays" for purge
```

Comment:
Update the comment to cover the new feature

Reply:
done